### PR TITLE
Revised rowwise() + condense()

### PR DESCRIPTION
**Pull request by romainfrancois**
Until we can commit to this new `do()`: 

``` r
library(dplyr, warn.conflicts = FALSE)
library(gapminder)

res <- gapminder %>% 
  group_by(continent, country) %>% 
  sketch(
    data = across(),
    mod = lm(lifeExp ~ year, data = data),
    pred = predict(mod, data)
  )
res
#> Source: local data frame [142 x 5]
#> Groups: <by row>
#> 
#> # A tibble: 142 x 5
#> # Groups:   [142]
#>    continent country                  data              mod    pred      
#>    <fct>     <fct>                    <list>            <list> <list>    
#>  1 Africa    Algeria                  <tibble [12 × 4]> <lm>   <dbl [12]>
#>  2 Africa    Angola                   <tibble [12 × 4]> <lm>   <dbl [12]>
#>  3 Africa    Benin                    <tibble [12 × 4]> <lm>   <dbl [12]>
#>  4 Africa    Botswana                 <tibble [12 × 4]> <lm>   <dbl [12]>
#>  5 Africa    Burkina Faso             <tibble [12 × 4]> <lm>   <dbl [12]>
#>  6 Africa    Burundi                  <tibble [12 × 4]> <lm>   <dbl [12]>
#>  7 Africa    Cameroon                 <tibble [12 × 4]> <lm>   <dbl [12]>
#>  8 Africa    Central African Republic <tibble [12 × 4]> <lm>   <dbl [12]>
#>  9 Africa    Chad                     <tibble [12 × 4]> <lm>   <dbl [12]>
#> 10 Africa    Comoros                  <tibble [12 × 4]> <lm>   <dbl [12]>
#> # … with 132 more rows
attr(res, "groups")
#> # A tibble: 5 x 2
#>   continent       .rows
#> * <fct>     <list<int>>
#> 1 Africa           [52]
#> 2 Americas         [25]
#> 3 Asia             [33]
#> 4 Europe           [30]
#> 5 Oceania           [2]
class(res)
#> [1] "rowwise_df" "grouped_df" "tbl_df"     "tbl"        "data.frame"
```

<sup>Created on 2020-01-16 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9000)</sup>

- it enlists everything
- returns its result as a grouped + rowwise data frame

### CODE REVIEW

**Comment by romainfrancois** [comment]
printing is off because there currently is no such thing as a grouped + rowwise data frame and this probably requires subtle care

**Comment by hadley** [comment]
Fixes #4723.

I think this is now ready for review. Sorry for hijacking your PR!

File: R/rowwise.r [hunk 1]
```diff
@@ -1,109 +1,114 @@
...
+#'  ~sim, ~n, ~mean, ~sd,
+#'     1,  1,     1,   1,
+#'     2,  2,     2,   4,
+#'     3,  3,    -1,   2
+#' )
+#' # Here I supply variables to preserve after the summary
+#' params %>%
+#'   rowwise(sim) %>%
+#'   summarise(z = rnorm(n, mean, sd))
+#'
```
Comment (romainfrancois) [inline R/rowwise.r]: Is this expected that this is still `rowwise [sim]` after the `summarise()` ?

``` r
library(dplyr, warn.conflicts = FALSE)

params <- tribble(
 ~sim, ~n, ~mean, ~sd,
    1,  1,     1,   1,
    2,  2,     2,   4,
    3,  3,    -1,   2
)

params %>%
  rowwise(sim) %>%
  summarise(z = rnorm(n, mean, sd))
#> # A tibble: 6 x 2
#> # Rowwise:  sim
#>     sim      z
#>   <dbl>  <dbl>
#> 1     1  1.75 
#> 2     2 -0.491
#> 3     2 -0.622
#> 4     3  0.289
#> 5     3 -3.94 
#> 6     3 -2.40

params %>%
  group_by(sim) %>%
  summarise(z = rnorm(n, mean, sd))
#> # A tibble: 6 x 2
#>     sim      z
#>   <dbl>  <dbl>
#> 1     1  1.66 
#> 2     2 -0.154
#> 3     2  5.35 
#> 4     3  0.676
#> 5     3 -1.15 
#> 6     3  2.87
```

File: R/summarise.R [hunk 2]
```diff
@@ -117,6 +117,12 @@ summarise.grouped_df <- function(.data, ...) {
   out
 }
 
+#' @export
+summarise.rowwise_df <- function(.data, ...) {
+  out <- NextMethod()
+  rowwise_df(out, group_vars(.data))
```
Comment (romainfrancois) [inline R/summarise.R]: Should this not lose one layer ?

**Comment by romainfrancois** [comment]
Should grouped and rowwise play together: 

``` r
library(dplyr, warn.conflicts = FALSE)

params <- tribble(
 ~g, ~sim, ~n, ~mean, ~sd, 
  1,   1,   1,     1,   1,
  1,   2,   2,     2,   4,
  2,   3,   3,    -1,   2
)
params %>% 
  group_by(g) %>% 
  rowwise(sim) %>% 
  summarise(z = rnorm(n, mean, sd))
#> # A tibble: 6 x 3
#> # Rowwise:  g, sim
#>       g   sim      z
#>   <dbl> <dbl>  <dbl>
#> 1     1     1  1.67 
#> 2     1     2  0.768
#> 3     1     2  1.13 
#> 4     2     3 -0.612
#> 5     2     3 -2.63 
#> 6     2     3  3.72
```

I would have expected that after the `summarise()` we would get back something *grouped by* g ?

**Comment by romainfrancois** [comment]
Or maybe something to snap out of rowwise mode, e.g. `unrowwise()` or perhaps a `group_by.rowwise_df()` method ?

``` r
library(dplyr, warn.conflicts = FALSE)

params <- tribble(
 ~g, ~sim, ~n, ~mean, ~sd, 
  1,   1,   1,     1,   1,
  1,   2,   2,     2,   4,
  2,   3,   3,    -1,   2
)

unrowwise <- function(data) {
  UseMethod("unrowwise")
}

unrowwise.data.frame <- function(data) {
  data
}

unrowwise.rowwise_df <- function(data) {
  vars <- group_vars(data)
  if (length(vars) > 1) {
    grouped_df(data, head(vars, -1))
  } else {
    as_tibble(data)
  }
}

# still rowwised by g and sim
params %>% 
  rowwise(g, sim) %>% 
  summarise(z = rnorm(n, mean, sd))
#> # A tibble: 6 x 3
#> # Rowwise:  g, sim
#>       g   sim      z
#>   <dbl> <dbl>  <dbl>
#> 1     1     1  1.63 
#> 2     1     2  1.78 
#> 3     1     2  0.694
#> 4     2     3  0.752
#> 5     2     3 -0.741
#> 6     2     3 -0.742

# unrowwise()
params %>% 
  rowwise(g, sim) %>% 
  summarise(z = rnorm(n, mean, sd)) %>% 
  unrowwise()
#> # A tibble: 6 x 3
#> # Groups:   g [2]
#>       g   sim      z
#>   <dbl> <dbl>  <dbl>
#> 1     1     1  0.249
#> 2     1     2 -2.70 
#> 3     1     2  0.329
#> 4     2     3 -0.950
#> 5     2     3 -2.26 
#> 6     2     3  0.123
```

<sup>Created on 2020-01-17 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9000)</sup>

File: R/condense.r [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+#' @export
+condense.data.frame <- function(.data, ...) {
+  cols <- condense_cols(.data, ...)
+
+  res <- group_keys(.data)
+  res[names(cols)] <- cols
+  res
+}
```
[the remark this answers is not in the document]
Comment (hadley) [reply R/condense.r]: Ooops, thanks

File: R/rowwise.r [hunk 4]
```diff
@@ -1,109 +1,120 @@
...
-rowwise <- function(data) {
-  abort_if_not(is.data.frame(data))
-
-  structure(data, class = c("rowwise_df", "tbl_df", "tbl", "data.frame"))
+#' df <- tibble(x = runif(6), y = runif(6))
+#' # Compute the mean of x and y in each row
+#' df %>% rowwise() %>% mutate(z = mean(c(x, y)))
+#'
+#' # Compute the minimum of x and y in each row
+#' df %>% rowwise() %>% mutate(z = mean(c(x, y)))
```
Comment (romainfrancois) [inline R/rowwise.r]:
```suggestion
#' df %>% rowwise() %>% mutate(z = min(c(x, y)))
```

Comment (romainfrancois) [reply R/rowwise.r]: Sorry I did not realize the examples were supposed to be different and `mean()` was a typo for `min()`

**Comment by hadley** [comment]
I can't reproduce your output — the rowwise-ness is always preserved for me.  I added a couple of lines to the docs about how rowwise status is always preserved and what do use to get out of it. I'd prefer to not add another verb for this, and instead tell people to use `group_by()` or `as_tibble()`.

Since I think you'll use  rowwise `summarise()` primarily to produce more rows, that selectively ungrouping will be confusing, and I think the principle that the rowwise property is never lost makes things simpler (esp since the group dropping of `summarise()` causes a lot of confusion).

**Comment by romainfrancois** [comment]
That works fine as a principle for `summarise(<rowwise_df>)`. This makes `rowwise()` much more compelling than before.

File: R/rowwise.r (hunk 4 above)
Comment (hadley) [reply R/rowwise.r]: That's what I figured 😄

**Pull request merged**